service/shutdown.go

```diff
@@ -0,0 +1,56 @@
+// Package service provides functionality common to services
+package service
+
+import (
+	"context"
+	"errors"
+	"time"
+
+	"github.com/sourcegraph/conc/pool"
```

Comment:
When we are dealing with Golang concurrency is safe to use third libraries that abstract some concepts like this one?

Reply:
Depends on the library and what you are doing :-). I started using conc for a while and so far didn't suffered from any bugs (code looks pretty good overall too). But there is always a chance the library will screw you..so if it is something very simple prefer the stdlib, but for something kinda like a concurrent map/reduce this library makes it fairly easier.

Reply:
Also..for the sake of learning..when playing around on personal project I would advocate using only the stdlib/waitgroups/channels/mutexes so you can learn the basic building blocks. But more high level ones can be useful, golang/x has something similar with errgroup: https://pkg.go.dev/golang.org/x/sync/errgroup, but conc is more expressive since it leverages parametric typing.

Some stuff on conc I find quite complex/cumbersome..not sure I would use everything..but I do like the Pools.